File: rerun_cpp/tests/archetypes/arrows3d.cpp
#include "archetype_test.hpp"

#include <rerun/archetypes/arrows3d.hpp>

using namespace rerun::archetypes;

#define TEST_TAG "[arrow3d][archetypes]"

SCENARIO(
    "Arrows3D archetype can be serialized with the same result for manually built instances and "
    "the builder pattern",
    TEST_TAG
) {
    GIVEN("Constructed from builder and manually") {
        auto from_builder = Arrows3D::from_vectors({{1.0, 2.0, 3.0}, {10.0, 20.0, 30.0}})
                                .with_origins({{4.0, 5.0, 6.0}, {40.0, 50.0, 60.0}})
                                .with_radii({1.0, 10.0})
                                .with_colors({{0xAA, 0x00, 0x00, 0xCC}, {0x00, 0xBB, 0x00, 0xDD}})
                                .with_labels({"hello", "friend"})
                                .with_class_ids({126, 127});

        Arrows3D from_manual;
        from_manual.vectors = rerun::ComponentBatch::from_loggable<rerun::components::Vector3D>(
                                  {{1.0, 2.0, 3.0}, {10.0, 20.0, 30.0}},
                                  Arrows3D::Descriptor_vectors
        )
                                  .value_or_throw();
        from_manual.origins = rerun::ComponentBatch::from_loggable<rerun::components::Position3D>(
                                  {{4.0, 5.0, 6.0}, {40.0, 50.0, 60.0}},
                                  Arrows3D::Descriptor_origins
        )
                                  .value_or_throw();
        from_manual.radii = rerun::ComponentBatch::from_loggable<rerun::components::Radius>(
                                {1.0, 10.0},
                                Arrows3D::Descriptor_radii
        )
                                .value_or_throw();
        from_manual.colors = rerun::ComponentBatch::from_loggable<rerun::components::Color>(
                                 {{0xAA, 0x00, 0x00, 0xCC}, {0x00, 0xBB, 0x00, 0xDD}},
                                 Arrows3D::Descriptor_colors
        )
                                 .value_or_throw();
        from_manual.labels = rerun::ComponentBatch::from_loggable<rerun::components::Text>(
                                 {"hello", "friend"},
                                 Arrows3D::Descriptor_labels
        )
                                 .value_or_throw();
        from_manual.class_ids = rerun::ComponentBatch::from_loggable<rerun::components::ClassId>(
                                    {126, 127},
                                    Arrows3D::Descriptor_class_ids
        )
                                    .value_or_throw();

        test_compare_archetype_serialization(from_manual, from_builder);
    }
}

File: crates/store/re_sdk_types/tests/types/arrows3d.rs
use re_sdk_types::archetypes::Arrows3D;
use re_sdk_types::components::{ClassId, Color, Position3D, Radius, ShowLabels, Text, Vector3D};
use re_sdk_types::datatypes::Vec3D;
use re_sdk_types::{Archetype as _, AsComponents as _, ComponentBatch as _};

#[test]
fn roundtrip() {
    let expected = Arrows3D {
        vectors: vec![
            Vector3D(Vec3D([1.0, 2.0, 3.0])),
            Vector3D(Vec3D([10.0, 20.0, 30.0])),
        ]
        .serialized(Arrows3D::descriptor_vectors()),
        origins: vec![
            Position3D(Vec3D([4.0, 5.0, 6.0])),    //
            Position3D(Vec3D([40.0, 50.0, 60.0])), //
        ]
        .serialized(Arrows3D::descriptor_origins()),
        radii: vec![
            Radius::from(1.0), //
            Radius::from(10.0),
        ]
        .serialized(Arrows3D::descriptor_radii()),
        colors: vec![
            Color::from_unmultiplied_rgba(0xAA, 0x00, 0x00, 0xCC), //
            Color::from_unmultiplied_rgba(0x00, 0xBB, 0x00, 0xDD),
        ]
        .serialized(Arrows3D::descriptor_colors()),
        labels: vec![
            Text::from("hello"),  //
            Text::from("friend"), //
        ]
        .serialized(Arrows3D::descriptor_labels()),
        class_ids: vec![
            ClassId::from(126), //
            ClassId::from(127), //
        ]
        .serialized(Arrows3D::descriptor_class_ids()),
        show_labels: ShowLabels(true.into()).serialized(Arrows3D::descriptor_show_labels()),
    };

    let arch = Arrows3D::from_vectors([[1.0, 2.0, 3.0], [10.0, 20.0, 30.0]])
        .with_origins([[4.0, 5.0, 6.0], [40.0, 50.0, 60.0]])
        .with_radii([1.0, 10.0])
        .with_colors([0xAA0000CC, 0x00BB00DD])
        .with_labels(["hello", "friend"])
        .with_class_ids([126, 127])
        .with_show_labels(true);
    similar_asserts::assert_eq!(expected, arch);

    eprintln!("arch = {arch:#?}");
    let serialized = arch.to_arrow().unwrap();
    for (field, array) in &serialized {
        // NOTE: Keep those around please, very useful when debugging.
        // eprintln!("field = {field:#?}");
        // eprintln!("array = {array:#?}");
        eprintln!("{} = {array:#?}", field.name());
    }

    let deserialized = Arrows3D::from_arrow(serialized).unwrap();
    similar_asserts::assert_eq!(expected, deserialized);
}
